npm/vite-plugin-cypress-esm/src/index.ts

```diff
@@ -0,0 +1,123 @@
+import debugFn from 'debug'
+import type { Plugin } from 'vite'
+import fs from 'fs'
+import path from 'path'
+const debug = debugFn('cypress:vite-plugin-mock-esm')
+
+const MODULE_IMPORTER_IDENTIFIER = '__cypressModule'
+const MODULE_DYNAMIC_IMPORTER_IDENTIFIER = '__cypressDynamicModule'
+
+const MODULE_CACHE_FILEPATH = path.resolve(__dirname, '../client/moduleCache.js')
+
+export const CypressEsm = (): Plugin => {
+  /**
+   * Remap static import calls to use the Cypress module cache
+   *
+   * ```
+   * import DefaultExport, { NamedExport, Other as Alias } from 'module'
+   * ```
+   *
+   * becomes
+   *
+   * ```
+   * import * as _cypress_module_1 from 'module';
+   * const DefaultExport = __cypressModule(_cypress_module_1);
+   * const { NamedExport, Other: Alias } = __cypressModule(_cypress_module_1);
+   * ```
+   */
+  const mapImportsToCache = (moduleId: string, code: string) => {
+    debug(`Remapping imports for module ${moduleId}`)
+    let counter = 0
+
+    const moduleIdentifier = moduleId.replace(/[^a-zA-Z\d]/g, '_')
+
+    const importRegex = /import (.+?) from (.*)/g
+
+    return code.replace(
+      importRegex,
+      (match, importVars: string, importTarget: string) => {
+        debug(`Mapping import ${counter + 1} in module ${moduleId}`)
+
+        let replacement = `import * as cypress_${moduleIdentifier}_${++counter} from ${importTarget};`
```

Comment:
It hadn't really registered with me that doing an `import *` would destroy any tree-shaking, but I guess there isn't much of that done in dev mode so maybe not a huge impact?

Reply:
I have no idea about perf yet, we should definitely do some benchmarks once we get all the bugs sorted out.